### PR TITLE
`reshape_input`, `reshape_output`

### LINKED ISSUE #895: what should be the returned shape of `Scattering1D` paths under `out_type='dict'` ?
Spotted while reviewing #871 with @janden

Right now, `out_type='dict'`behaves strangely. 

```python
J, Q = 2, 1
length = 1024
S = Scattering1D(J, shape, Q)
x = torch.zeros((5, 10,) + shape)

Sx = S(x)
print(Sx.shape)

S.out_type = "dict"
Sx = S(x)
print(list(map(lambda v: v.shape, Sx.values())))
```

Result: 
```
torch.Size([5, 10, 4, 256])

[torch.Size([5, 10, 1, 256]),
 torch.Size([5, 10, 1, 256]),
 torch.Size([5, 10, 1, 256]),
 torch.Size([5, 10, 1, 256])]
```

In other words, `out_type="dict"` is inserting a singleton dimension in penultimate position. This penultimate position serves to concatenate coefficients under `out_type="array"` (it becomes `4` in the example above).

I am in favor of removing this extra singleton dimension, and making values in `out_type="dict"` more like `'coef'`'s of `out_type="list"`. Compare with 

```python
S.out_type = "list"
Sx = S(x)
list(map(lambda v: v['coef'].shape, Sx))
```

Result:

```
[torch.Size([5, 10, 256]),
 torch.Size([5, 10, 256]),
 torch.Size([5, 10, 256]),
 torch.Size([5, 10, 256])]
```

This would be a breaking change, which is why i didn't make it single-handedly in my open PR #871.
I'm curious to hear everyone's thought here? bug or feature?

### PR DESCRIPTION
**Pull request by lostanlen**
A proposed fix for #870 

(EDIT: also fixes #895 )

includes a working demo in `Scattering1D` module
the same idea could work in other dimensions with the same methods `reshape_input` and `reshape_output` (no need to specialize them per 1D/2D/3D scattering!)

to help reviewers, i haven't moved the code from specialized frontends to the base frontend, only created call sites for `reshape_input` and `reshape_output` in 1D. But that's the idea of course. (could be done in a separate PR after review and resolution of merge conflicts)

### CODE REVIEW

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 1]
```diff
@@ -67,36 +67,26 @@ def load_filters(self):
                     n += 1
 
     def scattering(self, x):
+        self.load_filters()
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/torch_frontend.py:70]: why called before checks?

**Review by MuawizChaudhary — COMMENTED** [review]
I feel like its better to have these operations in the frontend rather than implemented in the backend. perhaps each classes base frontend class?

File: kymatio/scattering1d/frontend/torch_frontend.py (hunk 1 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/torch_frontend.py:70]: good question! because the goal is for this method to become:

```
def scattering(self, x):
    self.load_filters()
    super(ScatteringBase1D, self).scattering(x)
```

**Comment by lostanlen** [comment]
yes, we could do that, but consider that:

(1) padding already belongs to backends, so i don't see why other array ops like reshapes should be in frontends

(2) if we're putting it in the frontend, users will see it. and i don't think they'll have much use of methods called `scattering.reshape_input` and `scattering.reshape_output` . So we might turn them into private methods `scattering._reshape_input` and `scattering._reshape_output`. But i believe that it's preferable to have public methods in the backend module rather than private methods in the frontend object.

**Comment by lostanlen** [comment]
i'm having some difficulty rebasing this PR. it's not entirely clear to me what are the specifications of array shapes depending on whether we have out_type="array", "dict", or "list".

It seems like we insert a `1` between batch shape and signal shape that we end up removing sometimes, but not always? can anyone help?

**Comment by lostanlen** [comment]
alright, now attempting to unify the frontends.

the latest commit ([9637abb](https://github.com/kymatio/kymatio/pull/871/commits/9637abb65214936c48372d57ce019c0f1ef0b270)) is where the code diff is going to be easiest to understand

**Review by MuawizChaudhary — APPROVED** [review]
Pro merging

**Comment by lostanlen** [comment]
@MuawizChaudhary i'm assuming that this one too is being postponed to post-v0.3 ?

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 2]
```diff
@@ -18,38 +18,26 @@ def __init__(self, J, shape, Q=1, T=None, max_order=2, average=True,
...
-                new_shape = batch_shape + scattering_shape
-                x['coef'] = x['coef'].reshape(new_shape)
-
-        return S
+        n_kept_dims = 1 + (self.out_type=="dict")
+        for n, path in enumerate(S):
+            S[n]['coef'] = self.backend.reshape_output(
+                path['coef'], batch_shape, n_kept_dims=n_kept_dims)
+
+        if self.out_type=='array':
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/numpy_frontend.py:35]: `self.out_type == 'array'` please. same for others

File: kymatio/scattering1d/frontend/torch_frontend.py [hunk 3]
```diff
@@ -67,41 +67,28 @@ def load_filters(self):
...
-                new_shape = batch_shape + scattering_shape
-                x['coef'] = x['coef'].reshape(new_shape)
-
-        return S
+        n_kept_dims = 1 + (self.out_type=="dict")
+        for n, path in enumerate(S):
+            S[n]['coef'] = self.backend.reshape_output(
+                path['coef'], batch_shape, n_kept_dims=n_kept_dims)
+
+        if self.out_type=='array':
```
Comment (cyrusvahidi) [inline kymatio/scattering1d/frontend/torch_frontend.py:86]: same here `self.out_type == 'array'`

**Review by cyrusvahidi — CHANGES_REQUESTED** [review]
small edits, then LGTM

File: kymatio/scattering1d/frontend/torch_frontend.py (hunk 3 above)
Comment (lostanlen) [reply kymatio/scattering1d/frontend/torch_frontend.py]: 182ae3b

(search-and-replaced 12 times)

**Comment by MuawizChaudhary** [comment]
> @MuawizChaudhary i'm assuming that this one too is being postponed to post-v0.3 ?

I think that's the right decision here.

**Review by cyrusvahidi — APPROVED** [review]

File: kymatio/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -6,12 +6,26 @@ class TensorFlowBackend(NumpyBackend):
...
+        return tf.stack(arrays, axis=dim)
 
     @staticmethod
     def modulus(x):
         norm = tf.abs(x)
 
         return norm
 
+    @staticmethod
+    def reshape_input(x, signal_shape, n_inserted_dims=0):
```
Comment (janden) [inline kymatio/backend/tensorflow_backend.py:19]: Not following this `n_inserted_dims`, `n_kept_dims` business. Seems like the former is always `1` and the second is sometimes `1`, sometimes `2`. Should the latter case be handled by a `stack`–`concatenate` distinction?

**Review by janden — DISMISSED** [review]
Looks good but a question on these `dims` arguments.

File: kymatio/backend/tensorflow_backend.py (hunk 4 above)
Comment (lostanlen) [reply kymatio/backend/tensorflow_backend.py:19]: i have raised an issue https://github.com/kymatio/kymatio/issues/895

> Should the latter case be handled by a stack–concatenate distinction?

I did this to avoid introducing two kymatio primitives (stack and concatenate) when `reshape_output` does the job up to now. Although i agree that "stack" and "concatenate" are more recognizable in terms of what they do.
I am happy to make that distinction in the backends if you think it's helpful to understand the logic here.

Comment (janden) [reply kymatio/backend/tensorflow_backend.py:19]: > The former is always one for 1D. It would be two for 2D and three for 3D. Since i'm implementing this in a dimension-agnostic way, i need to pass `n_inserted_dims` as an input argument.

We're adding two singleton dimensions in 2D and three in 3D? Are you sure? If I understand correctly, this added dimension corresponds to the channel and is one for all the dimensions.

> For me, the extra `1` in `out_type="dict"` has no justification and we should present dict values as `(batch_size, time)`, just like with `"list"`.

I agree. We can discuss further on the issue.

Comment (lostanlen) [reply kymatio/backend/tensorflow_backend.py:19]: > We're adding two singleton dimensions in 2D and three in 3D? Are you sure? 

My bad. We're adding one singleton dimension no matter what the input dimension is.
2D code

```python
        batch_shape = input.shape[:-2]
        signal_shape = input.shape[-2:]
        input = input.reshape((-1,) + signal_shape)
        S = scattering( ...
        scattering_shape = S.shape[-3:]
        new_shape = batch_shape + scattering_shape
        S = S.reshape(new_shape)
```


So, `n_inserted_dims` doesn't need to be modifiable. I'll update accordingly.

Comment (lostanlen) [reply kymatio/backend/tensorflow_backend.py]: fixed in 4a274be

**Review by MuawizChaudhary — DISMISSED** [review]
I think we should leave this as a draft for now.

**Comment by MuawizChaudhary** [comment]
I'd be open after 0.3

**Comment by lostanlen** [comment]
@MuawizChaudhary but why ... ? this is backwards-compatible, does not introduce new feature, and is LOC-neutral. It can only make our life easier downstream ...

**Comment by lostanlen** [comment]
Rebased. While writing this i found out that our `concatenate`primitive has inconsistent default values for `dims` across backends, and that this kwarg should probably be called `axis` instead of `dims`, in the spirit of being compliant with the Array API.

File: kymatio/scattering1d/frontend/numpy_frontend.py [hunk 5]
```diff
@@ -36,7 +36,7 @@ def scattering(self, x):
         elif self.out_type == 'dict':
             return {path['n']: path['coef'] for path in S}
         elif self.out_type == 'list':
-            return list(map(lambda path: path.pop('n')), S)
+            return S
```
Comment (MuawizChaudhary) [inline kymatio/scattering1d/frontend/numpy_frontend.py:39]: i like this change.

**Comment by lostanlen** [comment]
@cyrusvahidi this is next in line. please merge when you're ready

**Review by cyrusvahidi — APPROVED** [review]
bon i've checked out the tests. LGTM
